Support derivatives and integrals in ApplyPointwise, rename to TransformVolumeData

nilsvu opened the pull request:
## Proposed changes

Apply python functions elementwise to all volume data in H5 files, including taking derivatives and volume integrals.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->
`spectre apply-pointwise` is now `spectre transform-volume-data` (or `spectre transform-vol`).
<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Define a Python type annotation for "any tensor"

nilsvu committed:
Support derivatives in ApplyPointwise

This adds support for elementwise kernels in ApplyPointwise, as well as
retrieving the mesh, coordinates, and Jacobians in kernels.
This means we can take derivatives over the grid.

nilsvu committed:
Fix use of `rstrip` in Python

`rstrip` removes trailing characters from the given set,
instead of the exact substring. Slicing is safer.

nilsvu committed:
Test ApplyPointwise CLI endpoint

nilsvu committed:
Support integrals in ApplyPointwise

nilsvu commented:
@geoffrey4444 FYI, here's the PR that will allow us to compute power monitors, Jacobian diagnostics, etc over volume data. It works like this:

```sh
spectre transform-vol H5FILES -d element_data -k spectre.domain.jacobian_diagnostic [--integrate]
```

The result of the function is written back into the H5 files and can be visualized in ParaView.

wthrowe commented:
Should this utility be renamed, since it does non-pointwise things now?

nilsvu commented:
@wthrowe If you're referring to derivatives: we do some derivatives in `src/PointwiseFunctions/` and I think that's ok. Derivatives are pointwise in the continuum limit, right :) If you're referring to integrals: that's more of a side product of this function, and you can read it as "apply this pointwise function, then also integrate it up", so seems fine to me too. Sounds ok?

nilsdeppe commented on src/DataStructures/Tensor/Python/__init__.py:
```diff
@@ -55,3 +55,10 @@ def __init__(self, inverse: bool):
 Jacobian = JacobianMeta(inverse=False)
 InverseJacobian = JacobianMeta(inverse=True)
+
+# Define a type annotation that means "any tensor". This should really be a
```
We do sort of have this in C++: TensorAsExpression. I'm not sure that's what you want here, but might be good to add why note in the comment.

nilsvu replied:
I need a common superclass of all Tensor types. `TensorAsExpression` isn't a superclass of `Tensor`, and also it has template parameters (I don't really understand its inheritance pattern).

nilsdeppe commented on src/Visualization/Python/ApplyPointwise.py:
```diff
@@ -463,9 +530,45 @@ def shift_magnitude(
+        integral_values = np.stack(list(integrals.values())).T
+        integral_names = list(integrals.keys())
+        if output:
+            if output.endswith(".txt"):
```
or `.dat` or `.tsv`?

nilsvu replied:
I just removed this check. Now the output is an H5 file if the file extension is `.h5`, and a text file otherwise.

nilsdeppe commented on src/Visualization/Python/ApplyPointwise.py:
```diff
@@ -244,6 +246,11 @@ def apply_pointwise(volfiles: Union[spectre_h5.H5Vol,
       kernels: List of pointwise transformations to apply to the volume data
         in the form of 'Kernel' objects.
+      integrate: Compute the volume integral over the kernels instead of
+        writing them back into the volume files.
```
Can you explicitly state what needs to be returned? I don't understand what happens if I return `\gamma_{ij}`. Do I get `\int d\xi J \gamma_{ij}` or `d\xi \gamma_{ij}`? Or can I only return scalars? Do I need to return the tensors multiplied by the Jacobian determinant?

nilsvu replied:
When you return `\gamma_{ij}` from the kernel then the integral that's computed is `\int d\xi J \gamma_{ij}`. That's one integral per component in inertial coords. Basically you write a kernel that returns a tensor and then decide "let's integrate it over the volume". I added some more docs.

nilsdeppe commented on src/Visualization/Python/ApplyPointwise.py:
```diff
@@ -62,11 +76,116 @@ def get_tensor_component_names(tensor_name: str,
+        default, the argument name is transformed to CamelCase to determine the
+        dataset name in the volume data file. Specify a mapping in
+        'map_input_names' to override the default.
+
```
Can you give an example of how one would for example compute the 1-index constraint in scalar wave? There's a lot to unpack here, and I think having an example would make that much easier to unpack.

How does this interact with C++ functions wrapped with pybind? Can you document that too? I'm not clear how I could call a function that takes `psi, mesh, inv_jac` and compute the 1-index constraint locally if I have a C++ bound function.

nilsvu replied:
Added some docs. To call a binding just pass something like `-k spectre.PointwiseFunctions.ScalarWave.one_index_constraint` to the CLI. Not sure how well that works for bindings that have multiple overloads, we'll see. Once a couple of suitable bindings are merged, like #4644 or #4637, I can add a test.

nilsdeppe replied:
Okay, sounds good!

nilsdeppe reviewed (changes requested):
You can squash. Mostly dox because I don't fully understand how I should be writing my python functions

wthrowe reviewed (changes requested):
No.  Let's name this correctly before it has lots of uses and we're stuck with it.

Derivatives are not pointwise, not even in some limit.  From what I can find in the code, the only person who has added derivatives to `PointwiseFunctions` is you.  The fact that I missed your error the first time doesn't mean I have to keep letting you do it.

I agree that integrals are a different sort of thing entirely, so I drop any objections I had related to them.

(Haven't reviewed the code itself.)

nilsvu commented:
@wthrowe ok let's talk about pointwise naming, maybe next Monday after the spectre call? Can we avoid holding up this PR with that?

wthrowe commented:
Three business days is not an unreasonably long code review.

nilsvu commented:
No the time is fine but if we want to rename the script I'd like to do it in a separate PR. I understand the change in this PR is related to the name, but renaming is easier in a separate PR.

nilsvu committed:
Delay a CLI error message

Trigger this error message later so it's easier to print
available subfiles before selecting a kernel.

nilsvu committed:
Parse pybind11 overloads

nilsvu committed:
Support DataVector arguments in ApplyPointwise

nilsvu commented:
Pushed three more commits extending support for pybind11 overloads and DataVector arguments (I'll add tests for the pybind11 overload parsing once we have some suitable bindings merged, like the jacobian diagnostic).

nilsdeppe reviewed (commented):
I have no requested changes on the new commits.

nilsvu committed:
Add test for parse_pybind11_signature

nilsvu committed:
Rename ApplyPointwise -> TransformVolumeData

nilsvu commented:
Rebased and renamed to `TransformVolumeData`

nilsvu commented:
@nikwit would you have some time to review this too?

wthrowe reviewed (dismissed):
I approve the naming changes.  I'm not good enough with python to really review the rest.

nikwit commented on src/Visualization/Python/ApplyPointwise.py:
```diff
@@ -133,52 +279,60 @@ def apply_pointwise(volfiles: Union[spectre_h5.H5Vol,
+                        for i, component in enumerate(transformed_tensor):
+                            transformed_tensor_data[
+                                i, element.data_slice] = component
+                    transformed_tensor = kernel.output_type(
```
here and other places this will do a copy even though this could be avoided for a lot of types, no?

nilsvu replied:
good call, I added `copy=False` so the Tensor constructor points into the Numpy array (this is supported through the buffer protocol, see bindings in Python/Tensor.cpp)

nikwit commented on src/Visualization/Python/ApplyPointwise.py:
```diff
@@ -63,49 +76,176 @@ def get_tensor_component_names(tensor_name: str,
+
+    if arg.annotation in Mesh.values():
+        return ElementArg("mesh")
+    elif (arg.name in ["logical_coords", "logical_coordinates", "xi"]
```
I think it would be better if you check what the name is and then assert that it has the right type

nilsvu replied:
Also removed the "xi" option because maybe that can be used to mean something other than logical coords (some people name the displacement field in elasticity "xi").

nikwit reviewed (changes requested):
Overall I like the idea of having this functionality and how most of it is structured. But I have some concerns about the using of typing and argument names for control flow in python. 

1.) I have never seen that the actual name of a function argument or variable in python (or any language) should affect the behavior of a program but here we are using it as keys to the observed quantities. Unless one reads the implementation (which takes a while to understand) it is a bit intransparent what is going on for a user. If we decide to keep it this way it should be pointed out most prominently in the documentation that this is happening.

2.) I don't like the mandatory typing in python. It is not a statically typed language and while typing hints can be a useful tool to make code a bit clearer I think they should be optional and not ever used for control flow ( they are called "hints" after all). I think it might even be possible to completely get rid of tensors on the python side and directly convert everything to appropriately sized `np.array`s. Then dynamic checking of the dims can be done. 

Overall, I would  prefer a simpler and more pythonic solution, such as 
`apply_pointwise(["Psi", "InertialCoordinates"], func_to_apply)` where `func_to_apply` takes `Psi` and `InertialCoordinates` as arguments. 
Is  there a reason why this is not possible?

nilsvu commented:
@nikwit The argument _names_ are used as default dataset names so you don't have to type them all on the command line (e.g. names like "shift", "lapse" etc are standard enough in our bound functions that it would be silly to write `-i shift=Shift` every time you invoke this script). To make this transparent to the user I added logging output that prints the dataset names that are used, so there's no magic going on behind the scenes (or at least, you understand it).

The argument _types_ are used to cast the data into the type that the function expects (mainly pybind11 bindings of C++ functions). I think that's a good use of type hints, considering that we're passing Python objects to C++, which is a statically typed language, and pybind11 actually asserts the argument types are correct when a binding is called. Maybe this can be done entirely by pybind11 though, using #4680. I'll play around with that a bit to see if it's enough to make the type hints optional.

nilsvu committed:
Support multiple outputs per kernel in TransformVolumeData

Also removes the requirement that kernels annotate their return value.

nilsvu commented:
@nikwit using #4680 I made return type annotations optional (please see last commit). You can now return tensors, DataVectors, or Numpy arrays from the kernels, and they get written to datasets in the volfiles. You can also return a dict from the kernel to specify the dataset names. We still need the argument type annotations so we can cast the dataset to the right tensor type. As I wrote above, I think that's a pretty good use of type annotations.
